samples/bluetooth/central_iso/src/main.c

```diff
@@ -19,64 +19,75 @@
 #include <zephyr/settings/settings.h>
 #include <zephyr/sys/byteorder.h>
 
+#define LATENCY_MS               10U                   /* 10.0 ms */
+#define INTERVAL_US              (10U * USEC_PER_MSEC) /* 10.0 ms */
+
 static void start_scan(void);
 
 static struct bt_conn *default_conn;
-static struct k_work_delayable iso_send_work;
 static struct bt_iso_chan iso_chan;
-static uint16_t seq_num;
-static uint16_t latency_ms = 10U; /* 10ms */
-static uint32_t interval_us = 10U * USEC_PER_MSEC; /* 10 ms */
 NET_BUF_POOL_FIXED_DEFINE(tx_pool, 1, BT_ISO_SDU_BUF_SIZE(CONFIG_BT_ISO_TX_MTU),
 			  CONFIG_BT_CONN_TX_USER_DATA_SIZE, NULL);
 
 /**
- * @brief Send ISO data on timeout
- *
- * This will send an increasing amount of ISO data, starting from 1 octet.
- *
- * First iteration : 0x00
- * Second iteration: 0x00 0x01
- * Third iteration : 0x00 0x01 0x02
- *
- * And so on, until it wraps around the configured ISO TX MTU (CONFIG_BT_ISO_TX_MTU)
+ * @brief Send ISO data SDU
  *
- * @param work Pointer to the work structure
+ * This will send a decreasing amount of ISO data until it reaches 1 byte then
+ * it wraps around.
  */
-static void iso_timer_timeout(struct k_work *work)
+static void iso_send_sdu(uint32_t seq_num, int32_t next_sdu_ts)
 {
-	int ret;
 	static uint8_t buf_data[CONFIG_BT_ISO_TX_MTU];
-	static bool data_initialized;
+	static size_t len_to_send = ARRAY_SIZE(buf_data);
+	bool is_first_sdu = (next_sdu_ts == -1);
 	struct net_buf *buf;
-	static size_t len_to_send = 1;
+	int ret;
 
-	if (!data_initialized) {
+	if (is_first_sdu) {
+		/* initialize mock data */
 		for (int i = 0; i < ARRAY_SIZE(buf_data); i++) {
 			buf_data[i] = (uint8_t)i;
 		}
-
-		data_initialized = true;
 	}
 
 	buf = net_buf_alloc(&tx_pool, K_FOREVER);
 	net_buf_reserve(buf, BT_ISO_CHAN_SEND_RESERVE);
-
 	net_buf_add_mem(buf, buf_data, len_to_send);
 
-	ret = bt_iso_chan_send(&iso_chan, buf, seq_num++);
+	if (unlikely(is_first_sdu)) {
+		ret = bt_iso_chan_send(&iso_chan, buf, seq_num);
+	} else {
+		ret = bt_iso_chan_send_ts(&iso_chan, buf, seq_num, next_sdu_ts);
+	}
 
 	if (ret < 0) {
 		printk("Failed to send ISO data (%d)\n", ret);
 		net_buf_unref(buf);
 	}
 
-	k_work_schedule(&iso_send_work, K_USEC(interval_us));
+	len_to_send--;
+	if (len_to_send == 0) {
+		len_to_send = ARRAY_SIZE(buf_data);
+	}
+}
 
-	len_to_send++;
-	if (len_to_send > ARRAY_SIZE(buf_data)) {
-		len_to_send = 1;
+static void iso_sent(struct bt_iso_chan *chan)
+{
+	struct bt_iso_tx_info iso_tx_info = {0};
+	int err;
+
+	err = bt_iso_chan_get_tx_sync(chan, &iso_tx_info);
```

Comment:
@fgrandel , thanks for spending time on improving this sample!
I would recommend to not use timestamps when providing data in this sample.
The increase in sequence number gives the controller enough information about when it is to be scheduled on air. See https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrfxlib/softdevice_controller/doc/isochronous_channels.html#providing-data for more information on how the Nordic SoftDevice Controller processing incoming HCI ISO Data.

When using timestamps it is possible to provide the data to the controller right before it is going to be sent on air.
Unfortunately, this timing information cannot be obtained in a controller-agnostic way:
The timestamp returned from the HCI LE Read ISO TX Sync command is ambiguous, the returned value depends on how the controller implementation interprets the spec. Also, the command documentation refers to CIG/BIG reference anchor points which may not align with the SDU generation if ISO_Interval != SDU interval (up to the controller). These things make it hard to rely on this return value.
See the open spec errata https://bluetooth.atlassian.net/browse/ES-23138 for more details.

Another note: The timing of the `iso_sent` is also implementation dependent. It is raised when the controller raises the HCI Number of Completed Packets event. This event may be raised when the controller is ready to receive more data. This may be the case when the payload is scheduled for transmission. That is, even before it is sent.

Reply:
TLDR; I see why the sent-callback + HCI LE Read ISO TX Sync + timestamp approach is lacking and depends on controller implementation details. But IMO it is still the only approach that works at all to reliably synchronize ISO channels for the moment being.

Arguments following:

Thanks for the link. Unfortunately I don't have permission to access it - guess that's only for SIG members.

The SDU vs. ISO interval argument is convincing - however as long as all our options are to some degree depending on the controller implementation, I'd still prefer the approach recommended by the core spec. By the time we'll implement more controllers, I guess the spec will have been updated and the errata removed in a somehow backwards compatible way, so we can move on from there.

In Zephyr we can and should decouple from implementation details by wrapping the controller specific parts behind more usable high-level APIs in the meantime. This would also improve developer experience by protecting them from the hacky and error-prone synchronization details. Plus it guarantees a migration path.

The dependence of iso_sent on the controller implementation is also a good argument. But we still need to provide developers with a working solution until the spec does so. That's AFAICS, why Nordic [recommends the HCI LE Read ISO TX Sync + timestamp solution](https://github.com/nrfconnect/sdk-nrf/blob/main/samples/bluetooth/iso_time_sync/src/iso_tx.c) for the moment being and uses the send callback as a trigger. Also see the recommended approach in the link you provided:

> This is the preferred way of providing data to the SoftDevice Controller and guarantees the highest degree of control.

Re sent-callback latency: Introducing a delay of one ISO interval w/o technical need is not a good way of interpreting the slack in the spec IMO. We should at least provide best-effort latency as far as technically possible - even more as it seems to be the only workable solution for deterministic synchronization for the time being. So the controller really should deliver some sufficient degree of latency guarantees beyond what is required by the spec for this specific callback/HCI command until the spec provides a workable solution itself.

The seq-number and "time-of-arrival" approaches don't work unless you admit at least two ISO intervals of extra latency to avoid systematic package loss (as demonstrated by the bug I'm fixing here). This I find inacceptable because it counters the concept of ISO channels with unframed PDUs and deterministic latency. Also see the latency promises of the CIG config API in Zephyr which will not be met. You need to add at least two sequence numbers all the time to ensure that:
1. the initially (but non-deterministically) skipped event gets accounted for.
2. source-to-CIG anchor point offset doesn't play a role even if you happen to synchronize too closely to the next anchor point so that the controller will pick up your SDUs too late.

In addition the "time-of-arrival" approach is not consistent with the spec AFAICS as flushing cannot be controlled deterministically by the application. It's the point of the ISO protocol that you can synchronize your source with the sink with deterministic latency and time-to-live of data packets.

On the peripheral both approaches don't work at all because neither the time-of-arrival nor the seq number approach are able to account for drift of the peripheral's clock. This cannot work by design if you want well-synchronized data across devices. This also explains why the (local) timer approach does systematically not work.

In our case for example we do decentralized measurements on peripherals in sync with the central's distributed clock across several devices and do explicitly want those measurements be sent out in the same CIG event with deterministic latency so we don't need timestamps and get cheap syntonization of all peripherals to the master clock w/o having to implement our own syntonization approach.

Reply:
> On the peripheral both approaches don't work at all because neither the time-of-arrival nor the seq number approach are able to account for drift of the peripheral's clock. This cannot work by design if you want well-synchronized data across devices. This also explains why the (local) timer approach does systematically not work.

Fun fact: An ISO peripheral from BT Core spec 5.4 is not aware of the SDU interval and can actually never transmit correctly (reliably) 8)

Reply:
Another comment for the use of `bt_iso_chan_get_tx_sync`: When we use it like this here, we can use the current response to compare with the previous response. IIRC (and please correct me @rugeGerritsen) the return values of `bt_iso_chan_get_tx_sync` are only when an SDUs have been sent (or scheduled for transmission). In the case that the timestamp is too low or the sequence number is our of order, the return value of the current and previous should be the same. 

e.g.

1) App send SDU_1 with TS_1 and PSN_1
2) App gets `sent` callback and gets TS_1 and PSN_1 from `bt_iso_chan_get_tx_sync`
3) App sends SDU_2 wit TS_1 and PSN_2 (or TS_2 and PSN_1)
4) Controller reject this as invalid (which not all controllers may do), but still triggers a Number of Completed Packets event
5) App gets `sent` callback and gets TS_1 and PSN_1 from `bt_iso_chan_get_tx_sync` (since it was not transmitted or scheduled for transmission)
6) App is now aware that the previous TX went wrong and could attempt to fix by increasing PSN or TS accordingly

Reply:
> In the case that the timestamp is too low or the sequence number is our of order, the return value of the current and previous should be the same.

Only when the Controller has received a SDU from Host, the returned Sequence Number and Timestamp correspond to the last transmitted or scheduled SDU are calculated.

1. If `bt_iso_chan_get_tx_sync` is called multiple times without a SDU enqueued, it will return same Sequence Number and Timestamp corresponding to the instant the previous `bt_iso_chan_send` was called.
2. Every `bt_iso_chan_get_tx_sync` after `bt_iso_chan_send` call will return the Sequence Number and Timestamp of the last transmitted or scheduled SDU. Hence, for the consecutive call to `bt_iso_chan_send`, the next Sequence Number (returned value plus one) and next Timestamp (returned value plus one SDU interval) be provided to `bt_iso_chan_send`.
3. If `bt_iso_chan_get_tx_sync` followed by `bt_iso_chan_send` is called in the sent-callback, then the provided Sequence Number and Timestamp will be stale depending on the SDU intervals elapsed from the last call of `bt_iso_chan_send` to when sent-callback executes.

Reply:
I've tried to implement a drift compensation algorithm now that:
* does no longer rely on the sent callback
* does not rely on the exact point in time that controllers store the TX sync information (even if the sync timestamp lies in the future)
* should continue to work if SDU interval != ISO interval
* uses an algorithm that is able to calculate the SDU interval even on a peripheral
* works with BT > 1 and FT > 1
* tolerates a certain amount of package loss
* compensates for initial non-standard delay of the second SDU (split controller ISOAL bug?)
* uses seq_num only when sending the sdu (no more timestamp dependency)

For obvious reasons I can not compensate for TX timestamp specification ambiguities but as I have to rely on hardware-specific timing anyway as long as the bluetooth clock is not exposed by the API [in some hardware-independent way](https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1597226731), this is not a problem in practice as the split controller defines the timestamp in the same way as the Nordic SD controller AFAICS. Let's hope that by the time more hardware is being supported by the sample, this ambiguity will have been fixed.

> An ISO peripheral from BT Core spec 5.4 is not aware of the SDU interval

Hrm, I think I found a way to derive the SDU interval from channel info that is available both, on the central and the peripheral:

Transport_Latency = CIG_Sync_Delay + FT x ISO_Interval - SDU_Interval, see Bluetooth Core, Version 5.4, Vol 6, Part G, Section 3.2.2.

Therefore:
SDU_Interval = CIG_Sync_Delay + FT x ISO_Interval - Transport_Latency

These parameters are available from the HCI LE CIS Established event.